File: 2019/day04-1.ts
function solution(content: string) {
  return generate(content)
    .map(String)
    .filter((num) =>
      num
        .split("")
        .every(
          (item, index, array) =>
            index >= array.length - 1 || item <= array[index + 1]
        )
    )
    .filter((num) =>
      num
        .split("")
        .some(
          (item, index, array) =>
            index < array.length - 1 && item === array[index + 1]
        )
    ).length;
}

function generate(content: string) {
  const [lo, hi] = content.split("-").map(Number);
  return Array.from({ length: hi - lo }, (_, index) => lo + index);
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2019/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    lo, hi = [int(x) for x in content.split("-")]
    return len(
        [
            number
            for number in (str(x) for x in range(lo, hi))
            if all(a <= b for a, b in zip(number, number[1:]))
            and any(a == b for a, b in zip(number, number[1:]))
        ]
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
